test_workspace/recordset.bosatsu

```diff
@@ -0,0 +1,61 @@
+package RecordSet
+
+enum RecordValue:
+  RVBool(value: Bool)
+  RVInt(value: Int)
+  RVString(value: String)
+struct Field[t](name: String, toValue: t -> RecordValue)
+struct RecordSet[shape](fields: shape[Field[String], Field[Int], Field[Bool]], rows: List[shape[String,Int, Bool]])
+
+def restructure(RecordSet(fields, rows), f: forall a,b,c. shape1[a, b, c] -> shape2[a, b, c]) -> RecordSet[shape2]:
+  RecordSet(f(fields), rows.map_List(f))
+
+struct Source[a,b,c](s1: a, s2: b, s3: c)
+struct Target[a,b,c](t1: a, t2: c, t3: b)
+
+rs = RecordSet(
+  Source(Field("String", RVString), Field("Int", RVInt), Field("Bool", RVBool)),
+  [Source("a", 1, True)]
+)
+
+def swap_last_two(Source(first, second, third)):
+  Target(first, third, second)
+
+rs2 = rs.restructure(swap_last_two)
+
+def and(x, y):
+  match x:
```

Comment:
small note: `def and(x, y): y if x else False` is a one liner for this, but ultimately the same code.

Reply:
nice!